src/org/mozartoz/truffle/nodes/call/TailCallCatcherNode.java

```diff
@@ -0,0 +1,45 @@
+package org.mozartoz.truffle.nodes.call;
+
+import org.mozartoz.truffle.nodes.DerefNode;
+import org.mozartoz.truffle.nodes.OzNode;
+import org.mozartoz.truffle.runtime.OzArguments;
+import org.mozartoz.truffle.runtime.TailCallException;
+
+import com.oracle.truffle.api.dsl.CreateCast;
+import com.oracle.truffle.api.frame.VirtualFrame;
+import com.oracle.truffle.api.nodes.IndirectCallNode;
+
+public class TailCallCatcherNode extends OzNode {
+
+	@Child OzNode call;
+	@Child IndirectCallNode callNode = IndirectCallNode.create();
+
+	public TailCallCatcherNode(OzNode call) {
+		this.call = call;
+	}
+
+	@CreateCast("function")
+	protected OzNode derefFunction(OzNode var) {
+		return DerefNode.create(var);
+	}
+
+	@Override
+	public Object execute(VirtualFrame frame) {
+		TailCallException tailCall = null;
+		while (true) {
```

Comment:
There should be some profiling here, to make sure normal calls don't get overhead.